### PR TITLE
PLT-619 Refactor and rename ChainIndexTxOut

### PR DESCRIPTION
**Pull request by berewt**
It mostly relies on the proposal of PR #626.
It provides type safety for script / pubkey txout (storing either the script hash or the pubkey), an extra getter to get the address from a txout, and smart constructors to build txout from an address.
In the refactoring of other modules, I relied as much as possible on lenses to avoid direct access to the partial field accessors (I'd like to avoid exposing them, but they are used in record syntax constructors in a few places, and it makes sense to use them in this context).

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-ledger/src/Ledger/Tx.hs [hunk 1]
```diff
@@ -120,89 +134,116 @@ datumInDatumFromQuery _ DatumUnknown    = pure DatumUnknown
...
--- This datatype was created in order to be used in
--- 'Ledger.Constraints.processConstraint', specifically with the constraints
--- 'MustSpendPubKeyOutput' and 'MustSpendScriptOutput'.
-data ChainIndexTxOut =
-    PublicKeyChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- public key hash.
-      _ciTxOutAddress         :: Address,
+-- | Offchain view of a transaction output.
+data OffchainTxOut =
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:138]: Sorry for not saying it earlier...

May I suggest the initial name of the previous PR? `DecoratedTxOut`? My reasoning:
There is already `Cardano.Api.TxOut` which is an offchain `TxOut`. But the current `OffchainTxOut` is not exactly that. It is an offchain `TxOut` but with extra stuff inside. Hence why a name like `DecoractedTxOut`, `TxOutWithExtras`, `EnrichedTxOut`, etc. might be better.

File: plutus-contract/test/Spec/Contract/TxConstraints.hs [hunk 2]
```diff
@@ -0,0 +1,299 @@
+{-# LANGUAGE BlockArguments        #-}
```
Comment (koslambrou) [inline plutus-contract/test/Spec/Contract/TxConstraints.hs:1]: Why was this file added? Seems like an old file that was added back.. Am I wrong?

Comment (berewt) [reply plutus-contract/test/Spec/Contract/TxConstraints.hs:1]: I guess I didn't bay enough attention during the rebase (I probably changed something to it and git thought I wanted to keep it while James removed it).

File: plutus-ledger/src/Ledger/Tx.hs (hunk 1 above)
Comment (berewt) [reply plutus-ledger/src/Ledger/Tx.hs:138]: Let's go with `Decorated` then. If I'm fortunate enough my `sed` skills should be sufficient to change it quickly.

File: plutus-ledger/src/Ledger/Tx.hs [hunk 3]
```diff
@@ -120,89 +134,116 @@ datumInDatumFromQuery _ DatumUnknown    = pure DatumUnknown
...
+
+_offchainTxOutAddress :: OffchainTxOut -> Address
+_offchainTxOutAddress PublicKeyOffchainTxOut{_offchainTxOutPubKeyHash, _offchainTxOutStakingCredential} =
+    V1.Address (V1.PubKeyCredential _offchainTxOutPubKeyHash) _offchainTxOutStakingCredential
+_offchainTxOutAddress ScriptOffchainTxOut{_offchainTxOutValidatorHash, _offchainTxOutStakingCredential} =
+    V1.Address (V1.ScriptCredential _offchainTxOutValidatorHash) _offchainTxOutStakingCredential
+
+offchainTxOutAddress :: Getter OffchainTxOut Address
+offchainTxOutAddress = let
+    in to _offchainTxOutAddress
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:207]: Remove "let in"?
